# Ejercicio:

-Con la problemática anterior, construya una metodología la cual analice si realmente dicho modelo también sirve con datosque no fueron usados para ajustar el algoritmo.

Tips: Recuerda el análisis de overfitting y underfitting visto en clase.

Ahora, ya tenemos más herramientas de evaluación para analizar la effectividad del modelo.

-Use las métricas de evaluación vistas en la clase de hoy y anteriores, para sacar más conclusiones sobre la efectividad del modelo.

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('conversiones.csv')

In [3]:
data["Conversion"] = pd.Categorical(data["Conversion"]).codes

In [6]:
y = data["Conversion"]
X = data.drop(["Conversion"], axis = 1)

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix


In [32]:
#División de datos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Entrenamiento del modelo
knn_modelo = KNeighborsClassifier(n_neighbors=3)  # Se ajustó el hiperparámetro a 3
knn_modelo.fit(X_train, y_train)

#Predicción en el conjunto de testeo
y_pred = knn_modelo.predict(X_test)

#Evaluación del rendimiento con la exactitud
Exactitud = accuracy_score(y_test, y_pred)
print(f'Exactitud: {Exactitud:.2f}')

#Evaluación del rendimiento con Precisión
Precision = precision_score(y_test, y_pred, average= None)
print('Precisión:', Precision)

#Evaluación del rendimiento con Recall 
Sensibilidad = recall_score(y_test, y_pred, average= None)
print('Sensibilidad:',Sensibilidad)

#Evaluación del rendimiento con F1-Score
F1S = f1_score(y_test, y_pred, average= None)
print('F1Score:', F1S)

#Evaluación del rendimiento con Matriz de Confusión
Matriz = confusion_matrix(y_test, y_pred)
print('Matriz de Confusión:', Matriz)

Exactitud: 0.98
Precisión: [0.99294974 0.95859805]
Sensibilidad: [0.95694649 0.99322996]
F1Score: [0.97461573 0.97560677]
Matriz de Confusión: [[30562  1375]
 [  217 31836]]


In [27]:
len(y_pred)

63990

Intepretación de la matriz de Confusión 

El modelo clasificó correctamente 30562 instancias como negativas (TN).

El modelo cometió 1375 falsas alarmas, es decir, clasificó erróneamente 1,375 instancias como positivas cuando en realidad eran negativas (FP).

El modelo omitió 217 instancias que eran positivas y las clasificó incorrectamente como negativas (FN).

El modelo clasificó correctamente 31.836 instancias como positivas (TP).